pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ProducerImpl.java

```diff
@@ -484,7 +497,105 @@ void ackReceived(ClientCnx cnx, long sequenceId, long ledgerId, long entryId) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("[{}] [{}] Corrupt message is already timed out {}", topic, producerName, sequenceId);
+                    }
+                }
+            }
+        }
+        // close connection and let producer resend pending-messages
+        cnx.ctx().close();
+    }
+    
+    /**
+     * Computes checksum again and verifies it against existing checksum. 
+     * If checksum doesn't match then updates newly computed checksum {@link OpSendMsg}. 
+     * 
+     * @param op
+     * @return isUpdated: returns true only if checksum is updated in {@link OpSendMsg}
+     */
+    protected boolean updateChecksumIfRequire(OpSendMsg op) {
+        
+        if (op.cmd instanceof DoubleByteBuf) {
+            DoubleByteBuf msg = (DoubleByteBuf) op.cmd;
+            ByteBuf headerFrame = msg.getFirst();
+            msg.markReaderIndex();
+            headerFrame.markReaderIndex();
+            try {
+                // skip bytes up to checksum index
+                headerFrame.skipBytes(4); // skip [total-size]
+                int cmdSize = (int) headerFrame.readUnsignedInt();
+                headerFrame.skipBytes(cmdSize);
+                // verify if checksum present
+                int magicBytesIndex = headerFrame.readerIndex();
+                Long checksum = Commands.readChecksum(headerFrame);
+
+                if (checksum != null) {
+                    // msg.first is already at header-payload index, Recompute checksum for headers-payload
+                    long computedChecksum = computeChecksum(msg);
+                    if (checksum.intValue() != computedChecksum) {
+                        // set checksum write-index to write recomputed checksum:
+                        // checksumIndex = magicBytesIndex + 2bytes
+                        headerFrame.setInt(magicBytesIndex + 2, (int) computedChecksum);
+                        if (log.isDebugEnabled()) {
+                            log.debug("[{}] [{}] message is updated with new checksum {}", topic, producerName,
+                                    computedChecksum);
+                        }
+                        return true;
+                    } else {
+                        if (log.isDebugEnabled()) {
+                            log.debug("[{}] [{}] message is already corrupted with checksum {}", topic, producerName,
+                                    computedChecksum);
+                        }
+                    }
+                } else {
+                    log.warn("[{}] [{}] checksum is not present into message with id {}", topic, producerName,
+                            op.sequenceId);
+                }
+            } finally {
+                headerFrame.resetReaderIndex();
+                msg.resetReaderIndex();
+            }
+        } else {
+            log.warn("[{}] [{}] Memory leak detected while creating message with id {}", topic, producerName,
```

Comment:
It seems the very first object of DoubleByteBuf is created as ```SimpleLeakAwareByteBuf```  or ```AdvancedLeakAwareByteBuf``` based on ```ResourceLeak Level```. And it can be reproduced by starting any test-case and put the break-point at [Commands](https://github.com/yahoo/pulsar/blob/master/pulsar-common/src/main/java/com/yahoo/pulsar/common/api/Commands.java#L443) where we create ```DoubleByteBuf```. and at very first time it will create ```LeakAwareByteBuf```.

It is because of following code which [netty](https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/ResourceLeakDetector.java#L224) has:

```
         if ((leakCheckCnt ++ & mask) == 0) {
                reportLeak(level);
                return new DefaultResourceLeak(obj);
            } else {
                return null;
            }
```
 At very first time ```leakCheckCnt``` will be 0 and it reports leak and it cause to create ```LeakAwareByteBuf```